src/ILCompiler.Compiler/src/Compiler/ExportedMethodsRootProvider.cs

```diff
@@ -36,11 +45,45 @@ public void AddCompilationRoots(IRootingServiceProvider rootProvider)
                     if (ecmaMethod.IsNativeCallable)
                     {
                         string nativeCallableExportName = ecmaMethod.GetNativeCallableExportName();
+
                         if (nativeCallableExportName != null)
+                        {
+                            _methods.Add(ecmaMethod);
                             rootProvider.AddCompilationRoot(method, "Native callable", nativeCallableExportName);
+                        }
                     }
                 }
             }
+
+            // Only emit exported methods for user supplied assemblies
+            // _exportsFile is null when we're not doing a shared library build
+            if (_module != _module.Context.SystemModule && _exportsFile != null)
+                EmitExportedMethods();
+        }
+
+        private void EmitExportedMethods()
+        {
+            string moduleName = Path.GetFileNameWithoutExtension(_exportsFile);
+            StringBuilder stringBuilder = new StringBuilder();
```

Comment:
Nit: It is better to write to the stream directly as you go instead of creating the whole string in memory and the writing it at once. (It does not matter a lot here since the file will be relatively small.)

```
            using (var writer = new StreamWriter(fileName))
            {
                writer.Write(...);
                writer.WriteLine(...);
            }
```